tsl/src/compression/api.c

```diff
@@ -396,8 +396,16 @@ decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_
 	/* Recreate FK constraints, since they were dropped during compression. */
 	ts_chunk_create_fks(uncompressed_chunk);
 
-	/* Prevent readers from using the compressed chunk that is going to be deleted */
-	LockRelationOid(uncompressed_chunk->table_id, AccessExclusiveLock);
+	/*
+	 * Prevents readers from using the compressed chunk that is going to be
+	 * deleted. Calling performMultipleDeletions in chunk_index_tuple_delete
+	 * also requests an AccessExclusiveLock.
+	 *
+	 * However, this call makes the lock on the chunk explicit
+	 * (as done in PostgreSQL when tables are dropped,
+	 * see RemoveRelations).
+	 */
+	LockRelationOid(compressed_chunk->table_id, AccessExclusiveLock);
```

Comment:
could you elaborate why this is needed ? doesn't ts_chunk_drop (or a function called by it )acquire the AccessExclusiveLock before dropping the table?

Reply:
My understanding is that no `AccessExclusiveLock` on the compressed chunk is requested in our existing code before it is deleted. 

I want to be compliant with the way PostgreSQL is performing locks before tables are deleted and `performMultipleDeletions` is called. Maybe I am too overcautious at this point and we can trust the internals of `performMultipleDeletions`. Then the lock would be superfluous. 

Attached are the locks that are active before `performMultipleDeletions` is called in `ts_chunk_index_delete_by_chunk_id` and the `AccessExclusiveLock` is implicitly created on the chunk (`_timescaledb_internal.compress_hyper_2_7296_chunk` is uncompressed in this example). 

```
SELECT relation::regclass as name, locktype, database, relation, pid, mode, granted, fastpath, waitstart FROM pg_locks WHERE relation::regclass::text LIKE '%chunk' ORDER BY relation, locktype, mode, granted;
                       name                        | locktype | database | relation |   pid   |       mode       | granted | fastpath | waitstart 
---------------------------------------------------+----------+----------+----------+---------+------------------+---------+----------+-----------
 _timescaledb_catalog.chunk                        | relation |    42698 |    42810 | 1113329 | RowExclusiveLock | t       | t        | 
 _timescaledb_internal._hyper_1_1_chunk            | relation |    42698 |    43372 | 1113329 | ExclusiveLock    | t       | f        | 
 _timescaledb_internal._hyper_1_1_chunk            | relation |    42698 |    43372 | 1113329 | ShareLock        | t       | f        | 
 _timescaledb_internal.compress_hyper_2_7296_chunk | relation |    42698 |    83658 | 1113329 | ExclusiveLock    | t       | f        | 
(4 rows)
```


```
timescaledb-2.9.0-dev.so!chunk_index_tuple_delete(TupleInfo * ti, void * data) (/home/jan/timescaledb/src/chunk_index.c:669)
timescaledb-2.9.0-dev.so!ts_scanner_scan(ScannerCtx * ctx) (/home/jan/timescaledb/src/scanner.c:451)
timescaledb-2.9.0-dev.so!chunk_index_scan(int indexid, ScanKeyData * scankey, int nkeys, tuple_found_func tuple_found, tuple_filter_func tuple_filter, void * data, LOCKMODE lockmode) (/home/jan/timescaledb/src/chunk_index.c:502)
timescaledb-2.9.0-dev.so!ts_chunk_index_delete_by_chunk_id(int32 chunk_id, _Bool drop_index) (/home/jan/timescaledb/src/chunk_index.c:779)
timescaledb-2.9.0-dev.so!chunk_tuple_delete(TupleInfo * ti, DropBehavior behavior, _Bool preserve_chunk_catalog_row) (/home/jan/timescaledb/src/chunk.c:2879)
timescaledb-2.9.0-dev.so!chunk_delete(ScanIterator * iterator, DropBehavior behavior, _Bool preserve_chunk_catalog_row) (/home/jan/timescaledb/src/chunk.c:2954)
timescaledb-2.9.0-dev.so!ts_chunk_delete_by_name_internal(const char * schema, const char * table, DropBehavior behavior, _Bool preserve_chunk_catalog_row) (/home/jan/timescaledb/src/chunk.c:2981)
timescaledb-2.9.0-dev.so!ts_chunk_delete_by_relid(Oid relid, DropBehavior behavior, _Bool preserve_chunk_catalog_row) (/home/jan/timescaledb/src/chunk.c:3002)
timescaledb-2.9.0-dev.so!ts_chunk_drop_internal(const Chunk * chunk, DropBehavior behavior, int32 log_level, _Bool preserve_catalog_row) (/home/jan/timescaledb/src/chunk.c:3669)
timescaledb-2.9.0-dev.so!ts_chunk_drop(const Chunk * chunk, DropBehavior behavior, int32 log_level) (/home/jan/timescaledb/src/chunk.c:3678)
timescaledb-tsl-2.9.0-dev.so!decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_relid, _Bool if_compressed) (/home/jan/timescaledb/tsl/src/compression/api.c:413)
timescaledb-tsl-2.9.0-dev.so!tsl_decompress_chunk(FunctionCallInfo fcinfo) (/home/jan/timescaledb/tsl/src/compression/api.c:645)
timescaledb-2.9.0-dev.so!ts_decompress_chunk(FunctionCallInfo fcinfo) (/home/jan/timescaledb/src/cross_module_fn.c:86)
ExecInterpExpr(ExprState * state, ExprContext * econtext, _Bool * isnull) (/home/jan/postgresql-sandbox/src/REL_14_2/src/backend/executor/execExprInterp.c:749)
ExecInterpExprStillValid(ExprState * state, ExprContext * econtext, _Bool * isNull) (/home/jan/postgresql-sandbox/src/REL_14_2/src/backend/executor/execExprInterp.c:1824)
ExecEvalExprSwitchContext(ExprState * state, ExprContext * econtext, _Bool * isNull) (/home/jan/postgresql-sandbox/src/REL_14_2/src/include/executor/executor.h:339)
ExecProject(ProjectionInfo * projInfo) (/home/jan/postgresql-sandbox/src/REL_14_2/src/include/executor/executor.h:373)
ExecScan(ScanState * node, ExecScanAccessMtd accessMtd, ExecScanRecheckMtd recheckMtd) (/home/jan/postgresql-sandbox/src/REL_14_2/src/backend/executor/execScan.c:238)
ExecFunctionScan(PlanState * pstate) (/home/jan/postgresql-sandbox/src/REL_14_2/src/backend/executor/nodeFunctionscan.c:270)
ExecProcNodeFirst(PlanState * node) (/home/jan/postgresql-sandbox/src/REL_14_2/src/backend/executor/execProcnode.c:463)
```

Reply:
Won't this give spurious errors when trying to read compressed tables? If any readers reach this point, it would behave roughly like this, which does not make sense since the data is available in the uncompressed chunk corresponding to this compressed chunk.
![4676-locking](https://user-images.githubusercontent.com/1175692/190386879-5b5f1a66-36ff-4bf1-b7ef-2fc20dfc9914.png)

What am I missing?

Reply:
@mkindahl Thank you for creating the drawing. It illustrates the problem well. Before I write a longer reply, I would like to make sure that I understand your concern correctly.

* Are you concerned that the `AccessExclusiveLock` should be taken later in the code so that readers can access the chunk as long as possible or
* are you concerned about this situation in general (i.e., a reader that waits on a `AccessShareLock` of a chunk and the chunk no longer exists after the lock is granted)?

Reply:
I'm concerned about the latter. In this situation, the reader should not even reach this point in the code.

If a decompression is ongoing, the compressed chunk will be decompressed and removed and the reader should be re-routed to the uncompressed chunk where the data will reside once the decompression is one. Then it can start reading from the chunk once the lock is granted. With this locking pattern, readers will experience errors if they are unlucky and race with the decompression job.

Reply:
@mkindahl In the most common code path (when using a hypertable with at least one index), the readers are already blocked by the `AccessExclusiveLock` on the index caused by the `reindex_relation` call after decompressing the chunk. However, it's a good point to also consider hypertables without any index and grab the lock after the compressed chunk is removed from the catalog to route such reads properly. I changed the PR accordingly.

Are you generally in favor of requesting an explicit `AccessExclusiveLock` on the chunk before we delete it? I am not sure if this is really necessary. I introduced it because PostgreSQL [explicitly requests](https://github.com/postgres/postgres/blob/b0a55e43299c4ea2a9a8c757f9c26352407d0ccc/src/backend/commands/tablecmds.c#L1335) an `AccessExclusiveLock` before a table is dropped and `performMultipleDeletions` is invoked, and I want to be consistent with the way PostgreSQL implements similar functionality. However, the preconditions for calling the `performMultipleDeletions` function don't seem to be explicitly defined. I am not sure what is the correct/best solution here. Maybe you have some advice for me.

Reply:
> Are you generally in favor of requesting an explicit `AccessExclusiveLock` on the chunk before we delete it? I am not sure if this is really necessary. I introduced it because PostgreSQL [explicitly requests](https://github.com/postgres/postgres/blob/b0a55e43299c4ea2a9a8c757f9c26352407d0ccc/src/backend/commands/tablecmds.c#L1335) an `AccessExclusiveLock` before a table is dropped and `performMultipleDeletions` is invoked, and I want to be consistent with the way PostgreSQL implements similar functionality. However, the preconditions for calling the `performMultipleDeletions` function don't seem to be explicitly defined. I am not sure what is the correct/best solution here. Maybe you have some advice for me.

If that is a precondition for calling `performMultipleDeletions` then we should do it, and in general you need to lock the table that you're deleting, but this is not the same as the problem described above.

My comment above was more a result from the fact that you are trying to solve a deadlock and in the comment say that you're locking here to prevent readers from reading the chunk. If you change this and readers do not normally reach this path, you need to update the comment so that it is accurate.

Reply:
@mkindahl Indeed, the comment was outdated and misleading. I have updated it in the current version of the PR.

The deadlock was introduced in a608d7db614c930213dee8d6a5e9d26a0259da61 by requesting a lock for the uncompressed chunk instead of the compressed one (`uncompressed_chunk->table_id` instead of `compressed_chunk->table_id`). This is changed in this PR and solves the deadlock. 

The PostgreSQL documentation does not mention whether such an `AccessExclusiveLock` is a prerequisite for calling `performMultipleDeletions`. But PostgreSQL explicitly requests such a lock before calling the function. I am therefore also in favor of explicitly requesting this lock.